### PR TITLE
List of strategic merge patches

### PR DESCRIPTION
**Pull request by hyww**
Support multiple strategic merge patches in a `List` like [this](https://github.com/kubernetes-sigs/kustomize/issues/638#issue-391605491).

Then we can put same patch with different targets in a  `List` and reuse duplicated parts using anchor/reference of YAML.

### CODE REVIEW

**Comment by Liujingfang1** [comment]
Can you add some tests?

**Comment by hyww** [comment]
@Liujingfang1 Sure. I have added some.

File: pkg/resource/factory.go [hunk 1]
```diff
@@ -66,7 +68,31 @@ func (rf *Factory) SliceFromPatches(
 		if err != nil {
 			return nil, internal.Handler(err, string(path))
 		}
-		result = append(result, res...)
+		for len(res) > 0 {
```
Comment (Liujingfang1) [inline pkg/resource/factory.go]: I suggest to put this change into `SliceFromBytes` and leave `SliceFromPatches` untouched. Then both resources and patches will be able to read `List`.

File: k8sdeps/kunstruct/factory_test.go [hunk 2]
```diff
@@ -33,6 +33,15 @@ func TestSliceFromBytes(t *testing.T) {
 				"name": "winnie",
 			},
 		})
+	testList := factory.FromMap(
```
Comment (Liujingfang1) [inline k8sdeps/kunstruct/factory_test.go]: Add a test to cover the usage in #638

File: k8sdeps/kunstruct/factory.go [hunk 3]
```diff
@@ -101,8 +101,9 @@ func (kf *KunstructuredFactoryImpl) Set(fs fs.FileSystem, ldr ifc.Loader) {
 }
 
 // validate validates that u has kind and name
+// except for kind `List`, which doesn't require a name
 func (kf *KunstructuredFactoryImpl) validate(u unstructured.Unstructured) error {
-	if u.GetName() == "" {
+	if u.GetName() == "" && u.GetKind() != "List" {
```
Comment (Liujingfang1) [inline k8sdeps/kunstruct/factory.go]: Can you also reorganize the if statements here? Put the checks for `Kind` together.
```
 	if u.GetName() == "" && u.GetKind() != "List" {
 		return fmt.Errorf("missing metadata.name in object %v", u)	 		
    }
	if u.GetKind() == "" {
 		return fmt.Errorf("missing kind in object %v", u)	 		 	
    }
```

Comment (hyww) [reply k8sdeps/kunstruct/factory.go]: Do you mean something like this?

note: this changed the priority of the two errors

```go
        kind := u.GetKind()
        if kind == "" {
                return fmt.Errorf("missing kind in object %v", u)
        } else if kind == "List" {
                return nil
        }
        if u.GetName() == "" {
                return fmt.Errorf("missing metadata.name in object %v", u)
        }
        return nil
```

Comment (Liujingfang1) [reply k8sdeps/kunstruct/factory.go]: Yep, this looks good.

**Comment by Liujingfang1** [comment]
/lgtm

Can you add a test for reading resources from a `List` in a follow up PR? Then we can mark #581 as fixed. Thank you.

**Pull request merged**